[WIP] Save step size et al for hmc and nuts

aseyboldt opened the pull request:
Right now, nuts and hmc do not keep step size, treedepth, acceptance rate... around. This patch saves them in `SamplerHist`.

Is is the right place for this data? The names `SamplerHist` and `state` seem a bit strange to me, but this is were the acceptance rate used to be stored. Also, why would it be possible to pass in an existing state object? I think I am missing something here...

Still missing: Nuts does still not store the acceptance rate and the energy. To get at the energy / the energy change between start end end point, I think I would need to change `buildtree`, and I don't really understand that well enough at the moment. And I'm a bit confused about how the acceptance is working now, too. Did the algorithm change recently?

aseyboldt committed:
Save step size et al for hmc and nuts

twiecki commented:
This is quite useful! Does it work with multiple chains?

ColCarroll commented:
@aseyboldt I could help with the energy change, but worry about speed/memory overhead.  I'll benchmark later so we're talking with with real numbers.

ColCarroll commented:
I actually just benchmarked this all with 100k samples, and it was pretty comparable to master for speed!

Two thoughts:
1. Would love to see utility functions that access the `state` to assess convergence, otherwise I worry this wouldn't be widely used (in fact, I removed the `state` variable a few months back since I did not realize there was interest)

2. Would love to see some tests :)

aseyboldt commented:
@twiecki No, I forgot about multiple chains. :-(
And this makes me think that the step-object is a weird place for this. Maybe it would be better to put it somewhere in the trace? That is how stan handles this. I guess we don't want to break the api of the backends, but maybe it could be an optional addition to this interface.

@ColCarroll I don't think this should have much of a performance impact on anything but the smallest models, the overhead does not depend on the number of variables.
I'm not sure what convergence checking would be appropriate here. But at least some form of warning about diverging transitions would be useful. Shinystan prints a big red warning about those, as they usually indicate that the posterior has high curvature and/or that the target acceptance rate is too small. And I think the usual Rhat statistic should work for the energy, too.
Point taken about the tests. :-)

I'll have a look if I can find a nice way to put this into the trace on Monday, if that is ok for you.

fonnesbeck commented:
Actually, this does most of what I'd be looking for, so I deleted my previous comment. Seems like it would make sense to add it to the model or trace as an attribute at runtime.

aseyboldt commented:
How about an api like this? This is only a sketch, still needs polishing, tests and doc if you agree that this is a good way to do this.

I added an attribute `generates_stats` to the base class of the samplers and a second attribute `stats_dtypes`. The base backends also gets a new attribute `supports_sampler_stats`. Both are False by default, but can be enabled by subclasses. The sampler then checks, if the step method exports stats and if the backend can store them. There is a bit of a complication because several step methods can be used at the same time, and they might export stats with the same name. To get around that problem I let the step methods export a list of dicts of stat values each time, that way a compound step method can just concatenate the lists of the individual step methods. A trace can then stack the values along a new axis. I included a minimal example notebook to show the api a bit.

fonnesbeck commented:
This looks suitable to me. One suggestion would be to make the individual statistics properties of the trace, so that you can access them easily. For example:

```
@property
def step_size_(self):
    return self.get_stats('step_size')
```

twiecki commented:
could we not store a second trace inside the sampler that just keeps track
of the sampler stats?

On Jan 23, 2017 2:57 PM, "Chris Fonnesbeck" <notifications@github.com>
wrote:

> This looks suitable to me. One suggestion would be to make the individual
> statistics properties of the trace, so that you can access them easily. For
> example:
>
> @property
> def step_size_(self):
>     return self.get_stats('step_size')
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/pymc3/pull/1687#issuecomment-274493970>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AApJmNJBKgg-7H9SzKY8YR4FK7abRmMuks5rVLG_gaJpZM4LpdZJ>
> .
>

fonnesbeck commented:
I'd worry that we'd start having to distinguish between MCMC output and diagnostic information if we treated them the same. Also, some of the information (like acceptance rate) is just a scalar.

aseyboldt commented:
@twiecki The logic for storing the stats and the logic for storing the variables is similar, but not the same. For example, we don't allow two variables with the same name. This can happen with the stats if we use multiple instances of the same step-method in a compound step method, however. And from a user point of view I also like the distinction between the actual trace and the diagnostics.

@fonnesbeck We can use the same mechanism for attribute access as for the trace variables (some stuff in __getattr__ that I commented out for now). If they share names, we could issue a warning and return the trace var.

twiecki commented on pymc3/backends/base.py:
```diff
@@ -213,6 +315,22 @@ def varnames(self):
         return self._straces[chain].varnames
 
+    @property
+    def statnames(self):
```
`stat_names`

fonnesbeck commented:
Would be great to get some fuller exposition in the notebook example. Ideally all of our notebooks should as much as possible be stand-alone, readable documents.

aseyboldt committed:
Fix tests for sampler stats

aseyboldt committed:
Improve notebook about sampler stats

aseyboldt commented:
I renamed `trace.statnames` to `trace.stat_names`, wrote a couple of additional tests (and fixed a bug in the sql backand along the way) and improved the notebook.

aseyboldt committed:
Fix tests for sampler stats

twiecki commented:
@aseyboldt Let us know when you want another round of review. The example NB looks great. Even better would be if it would say what good and bad behavior looks like and what to do about it.

aseyboldt commented:
Ready for review.

@twiecki I included a couple of sentences about good/bad behaviour, but doing this properly is not easy, and I'm not sure I'm the right person to do this. I'm still figuring a lot of this out for myself.

twiecki commented on pymc3/step_methods/hmc/nuts.py:
```diff
@@ -110,7 +130,17 @@ def astep(self, q0):
 
         self.m += 1
 
-        return q
```
We should also add the divergence -- the difference between the logp at the old and proposed position.

aseyboldt replied:
Do you mean the differences of the hamiltonians? I've got a half finished patch that does that and also adds the information if the transition was diverging and the maximal divergence along the trajectory. But I'd rather make that a separate pull request. It makes changes in the nuts code itself.

twiecki replied:
@aseyboldt I see, that sounds good to me.

twiecki commented on pymc3/backends/base.py:
```diff
@@ -105,6 +138,47 @@ def get_values(self, varname, burn=0, thin=1):
         """
         raise NotImplementedError
 
+    def get_stats(self, varname, sampler_idx=None, burn=0, thin=1):
```
I think I'd call this `get_sampler_stats()` as I'd assume `get_stats()` would get me stats of the posterior samples.

twiecki commented:
![image](https://cloud.githubusercontent.com/assets/674200/22619131/ff6eca38-eaed-11e6-92d3-2878bdfb7364.png)

I'd explicitly mention that the two chains are concatenated and that's why it looks weird. This motivates the next plot.

twiecki commented on pymc3/backends/base.py:
```diff
@@ -247,6 +370,39 @@ def get_values(self, varname, burn=0, thin=1, combine=True, chains=None,
             results = [self._straces[chains].get_values(varname, burn, thin)]
         return _squeeze_cat(results, combine, squeeze)
 
+    def get_stats(self, varname, burn=0, thin=1, combine=True, chains=None,
```
`get_sampler_stats`?

twiecki commented:
See minor comments but LGTM otherwise!

aseyboldt committed:
Rename backend.get_stats

aseyboldt committed:
Fix nuts failure after rebase

aseyboldt committed:
Update sampler-stats notebook

twiecki commented:
Thanks @aseyboldt!
